### PR TITLE
Improve PDB formatting with incomplete Monomer info

### PR DESCRIPTION
**Pull request by fwaibl**
# Introduction
By default, the PDB writer uses the residue name UNL, and sets very reasonable atom names:
```
>>> from rdkit import Chem
>>> mol = Chem.MolFromSmiles(r"ClCl")
>>> print(Chem.MolToPDBBlock(mol))
HETATM    1 CL1  UNL     1       0.000   0.000   0.000  1.00  0.00          CL  
HETATM    2 CL2  UNL     1       0.000   0.000   0.000  1.00  0.00          CL  
CONECT    1    2
END
```
We can also set the residue info manually. If this information is complete, everything works:
```
>>> at = mol.GetAtomWithIdx(0)
>>> res_inf = Chem.AtomPDBResidueInfo()
>>> res_inf.SetName("CLA ")
>>> res_inf.SetResidueName("CL2")
>>> res_inf.SetIsHeteroAtom(True)
>>> res_inf.SetResidueNumber(1)
>>> at.SetMonomerInfo(res_inf)
>>> print(Chem.MolToPDBBlock(mol))
HETATM    1 CLA  CL2     1       0.000   0.000   0.000  1.00  0.00          CL  
HETATM    2 CL1  UNL     1       0.000   0.000   0.000  1.00  0.00          CL  
CONECT    1    2
END
```
However, if the residue info is incomplete (e.g., we set the residue name but not the atom name), the PDB output will be invalid. Most importantly, the line is too short, so that the later columns get into the wrong place:
```
>>> mol = Chem.MolFromSmiles(r"ClCl")
>>> at = mol.GetAtomWithIdx(0)
>>> res_inf = Chem.AtomPDBResidueInfo()
>>> res_inf.SetResidueName("CL2")
>>> at.SetMonomerInfo(res_inf)
>>> print(Chem.MolToPDBBlock(mol))
ATOM      1  CL2     0       0.000   0.000   0.000  1.00  0.00          CL  
HETATM    2 CL1  UNL     1       0.000   0.000   0.000  1.00  0.00          CL  
CONECT    1    2
END
```
This happens if either the atom name or the residue name is missing. But IMO, the more important case is the missing atom name, since enumerating the atom names is more difficult that setting a dummy residue name.

# Changes
* This PR changes the code so that default atom names are generated if they are not present in the AtomPDBResidueInfo object. The same code is used as in the case without any residue info, and I refactored that code into a new function `GetDefaultAtomNumber`.
* I also added an `std::setw(4)` before the atom name and an `std::setw(3)` before the residue name, so that additional spaces will be printed before the atom or residue name if it is too short. (See below: I should still add a unit test for that). Again, this avoids incorrect column alignment. If the user specifies a short name like "C1", it will still not be aligned properly *within* the column, but the result will be parsed correctly by PyMol. (PDB specifies that the alignment of atom names depends on the length of the element symbol. To do that perfectly, we would have to parse the user-specified atom name).

# To do / Questions
- [x] I did not yet make a unit test where the atom name is set but the residue is not. I specified the minimal length in the code, but did not yet test if it works. I will add another commit for that.
- [ ] I added a function called `GetDefaultAtomNumber`, since that code is now used in two different places. I only declared and defined this function in the PDBWriter.cpp file, since I don't believe that it can be useful anywhere else, and since there is no corresponding .h file. Is that ok?
- [ ] What do you think of the alignment issue? I believe that adding spaces is an improvement, but it is still not perfect. However, the user can simply supply their atom names with the correct alignment.

What do you think of this? Thanks in advance for your help :-)

### CODE REVIEW

File: Code/GraphMol/FileParsers/PDBWriter.cpp
Comment (greglandrum) [inline Code/GraphMol/FileParsers/PDBWriter.cpp:141]: I think this would be better if you used either an `std::pair<char>` (a more natural container for something which is only two long) or an `std::string` (a more natural container for something which is inherently a string).

File: Code/GraphMol/FileParsers/file_parsers_catch.cpp
Comment (greglandrum) [inline Code/GraphMol/FileParsers/file_parsers_catch.cpp:3026]: constructing a molecule safely from an explicit SMILES is much simpler as:
```suggestion
      auto m = "Cl"_smiles;
```

This suggestion applies to all of the molecule construction below too.

File: Code/GraphMol/FileParsers/PDBWriter.cpp
Comment (greglandrum) [inline Code/GraphMol/FileParsers/PDBWriter.cpp:76]:
```suggestion
    if (name.empty()) {
```

File: Code/GraphMol/FileParsers/file_parsers_catch.cpp
Comment (greglandrum) [inline Code/GraphMol/FileParsers/file_parsers_catch.cpp:3023]: Since you've put the `setw()` in for the residue name, I think it would make sense to add a test for that here too: something that sets the residue name to a string with more than three characters and the verifies that the output PDB only includes the first 3 of those.

**Review by greglandrum — COMMENTED** [review]
Sorry for the very, very, very slow review.

I've requested a few changes here. Please let me know if you have questions about them.

**Review by greglandrum — CHANGES_REQUESTED** [review]
my intent with the original review was "request changes", not "comment".
sorry

**Comment by Jgmedina95** [comment]
wow, I was just dealing with this bug today. Thanks for trying to fix this, here I have a little workaround I did to do this less manually until the PR is merged for the case when you want to add the residue name only:
 ```
        m = Chem.MolFromSmiles(smiles_str)
        mol_name = "molecule_name"
        block = AllChem.MolToPDBBlock(m)
        for line in block.split("\n"):
            if line.startswith("HETATM"):  # avoiding CONECT, etc.
                atom_number = line[6:11]
                atom_name = line[12:16]
                res_number = line[22:26]
                atom = m.GetAtomWithIdx(int(atom_number) - 1)
                res_inf = Chem.AtomPDBResidueInfo()
                res_inf.SetName(f"{atom_name}")
                res_inf.SetResidueName(f"{code}")
                res_inf.SetResidueNumber(int(res_number))
                res_inf.SetIsHeteroAtom(True)
                atom.SetPDBResidueInfo(res_inf)
        if name:
            m.SetProp("_Name", name)

        Chem.MolToPDBFile(m, file_name)
```

I would love to help fix the bug, but my cpp knowledge is super super bad haha and I would probably brake it more than anything :D.

File: Code/GraphMol/FileParsers/PDBWriter.cpp
Comment (fwaibl) [reply Code/GraphMol/FileParsers/PDBWriter.cpp:76]: makes sense

File: Code/GraphMol/FileParsers/file_parsers_catch.cpp
Comment (fwaibl) [reply Code/GraphMol/FileParsers/file_parsers_catch.cpp]: Ok, I changed that

File: Code/GraphMol/FileParsers/PDBWriter.cpp
Comment (fwaibl) [reply Code/GraphMol/FileParsers/PDBWriter.cpp]: I changed it to be a std::string. That is also more convenient to work with in terms of output printing.

**Review by greglandrum — APPROVED** [review]
LGTM

**Comment by greglandrum** [comment]
I will merge this once the CI builds are done.

**Comment by greglandrum** [comment]
Thanks @fwaibl

**Pull request merged**
